Allow extensions to be built with MinGW in a virtualenv

rolk opened the pull request:
If using the setting `compiler=mingw32` (in for instance setup.cfg) when building an extension (with `python3 setup.py develop`) in a virtualenv, the build will end with the error message:

    Building import library (arch=AMD64): "XXX\libpython35.a" (from XXX\python35.dll)

where XXX is the path to your virtualenv, because the compiler adapter for MinGW searches only the sys.prefix for the libs/ directory containing the import libraries, and these are not copied to the virtualenv when it is created.

This raises the bar for compiling extensions with the MinGW distribution because a special extra step (copy the directory from the base distribution to the virtualenv) is now required, adding a burden on all extension developers to add this to their descriptions.

This patch adds sys.base_prefix to the search paths, to pick up the import libraries from the base distribution automatically. It is tested with WinPython 3.5.2.2 on Windows 7 (using MinGW 6.2.0 (win32, seh, rt-v5) toolset).

rolk committed:
BUG: Search for Python DLL in base distr. with MinGW

If an extension is built in a virtualenv, the compiler adapter for the
MinGW distribution will search for the libs/ directory containing the
import libraries for the Python runtime, in the virtualenv. However,
these files are not copied when the virtualenv is created (using the
venv module), so the build fails.

This patch searches the base distribution directory for the libraries
in addition to the virtualenv, allowing extensions to be built without
manually copying the libs/ directory first.

charris commented on numpy/distutils/mingw32ccompiler.py:
```diff
@@ -390,6 +391,13 @@ def _build_import_library_amd64():
+    base_file = os.path.join(sys.base_prefix, 'libs', out_name)
+    if os.path.isfile(base_file):
+        log.debug('Skip building import library: "%s" exists' %
+                  (base_file))
```
If you intend a tuple here, need a `,`, otherwise don't see the need for the parenthesis.

rolk replied:
Agreed, and upon inspection, string formatting should not be done in calls to log.debug in the first place. I'll submit a revised version. Some projects don't like pull requests to be rebased while in review, though; what is your stance on that?

charris replied:
We encourage cleaning up the commit history. PRs are considered private work areas before they are merged.

charris commented on numpy/distutils/mingw32ccompiler.py:
```diff
@@ -409,11 +417,22 @@ def _build_import_library_x86():
+        if os.path.isfile(base_lib):
+            lib_file = base_lib
+        else:
+            log.warn('Cannot build import library: "%s" not found' % (lib_file))
```
See above about parenthesis.

charris commented:
Makes sense to me, but not my area. @rgommers Thoughts?

rolk committed:
STY: Avoid using string formatting in log calls

The log class (derived from distutils.log) does string formatting with
extra arguments passed, if (and only if) the log level is above the set
threshold. That way, the cost of string formatting only incurs if the
message is actually going to be displayed.

charris commented:
OK, I've put your PRs in, lets hope they help. Thanks @rolk .
